draft/ZEP0010.md

```diff
@@ -0,0 +1,440 @@
+whereas ``attributes`` are generally intended for human interpretation and
+serve as passive metadata or provenance information, though the boundaries are
+not always distinct.
+
+By adding a new field, the specification can assert restrictions that if added
+to ``attributes``. would amount to a breaking change. If present, the
+``extensions`` field MUST contain an array of extension definitions. The
+contained array MUST either have one or more extensions or the object MUST be
+omitted entirely. Specifying metadata within ``extensions`` as opposed to
+``attributes`` allows the clear registration of the extension name, providing a
+namespace for the metadata to prevent collisions, and activates the
+``must_understand`` handling logic.
+
+Further details on the specification changes can be found in
+<https://github.com/zarr-developers/zarr-specs/pull/344>.
+
+### Definition and naming
+
+Each extension object will follow the rules laid out in the "Zarr extensions"
+section of the v3 specification.
+
+### Processing
+
+Zarr implementers
+are expected to inspect the extensions for each node and determine whether each listed
+extension is supported. If an extension includes ``"must_understand": true``
+(the default) and the implementation does not support it, the node must not
+be loaded and an appropriate error should be raised. For extensions with
+``"must_understand": false``, implementers may safely ignore unrecognized entries.
+
+To support a given extension, an implementation many hard-code a check for known
+extension names and invoke appropriate logic according to the extension’s
+specification at the correct point in its processing pipeline (e.g., during
+metadata interpretation, data access, or layout resolution).
+Where possible, however, implementations are encouraged, to delegate
+that logic via a callback or plugin mechanism that allows third-party code to
+handle the extension dynamically.
+
+As the set of extensions evolves, certain interfaces may arise which allow
+this modular approach for a subset of extensions. Where possible, these
+interfaces will be added to the specification. Feedback from implementers
+on such matters is highly encouraged.
+
+### Examples
+
+The following examples represent a few realistic use cases of the top-level
+``extensions`` container. This ZEP is putting in place the mechanism so the
+community can experiment with such extensions before their standardization.
+
+#### Offset (array)
+
+```javascript
+{
+    "zarr_format": 3,
+    "node_type": "array",
+    ...,
+    "extensions": [
+        {
+            "name": "example.offset",
+            "configuration": { "offset": [ 12, 24 ] }
+        }
+    ]
+}
+```
+
+The ``example.offset`` extension contains an array of the same order as the
+shape of the containing array specifying which element of the array should be
+considered as the origin, e.g., `[0, 0]`. This allows the reuse of subregions
+of an array without the need to rewrite the data.
```

Comment:
If someone were to come across a `must_understand=True` extension in a zarr.json in the wild, how would they find the information necessary for supporting it in their Zarr implementation?

Reply:
Is it possible to make this ZEP self-describing, so that a user knows where to find that information without needing to also read the Zarr V3 spec?